### DISCUSSION TITLE
Face Pose Angle Estimation

### DISCUSSION THREAD

**Discussion by saeidn95** [vladmandic/human#462] [Q&A]
I wonder if you can explain where the factor 2 for Theta values in this line comes from

https://github.com/vladmandic/human/blob/1114014bfd6c114baf2d31942382a2b2dc065c7d/src/face/angles.ts#L79

**Comment by vladmandic** [comment]
normalize to front view with 180drg range.

**Reply by saeidn95** [reply]
No sure I understand. Le'ts take yaw as example; As I understand if the heard is is not rotated, the yaw angle is zero. Turning the head 90 degrees left or right corresponds to +-90 degree with a range of 180 degrees. So, why do I need to multiply by 2?

**Reply by vladmandic** [reply]
math.asin has range -pi/2...pi/2 while math.atan2 has range -pi/pi

**Reply by saeidn95** [reply]
Well, I am not really sure of what you say; The formulation in (https://www.geometrictools.com/Documentation/EulerAngles.pdf) is scoped to work in the domain of (-pi/2 to +pi/2) where sin(.) is between (-1 to 1). That makes sense as none of the rotation around the head (yaw, pitch and roll) can go beyond -90 to +90. However, with slight addition to the formulation we can extend the rotation to full  (-pi to +pi). Let me explain. For example take the case of Section 2.6 with rotation order RzRyRx.

Here, thera_y = asin(-r20) only gives partial value  of thera_y wo within (-pi/2 to pi/2). But this result is only partial. We might have to add an offset +pi/2 or -pi/2. But how do I decide when I need to add this offset. That information comes from c_y that appear on the r00, r10, and r21 and r22. We use r00, r10 to obtain theta_z = atan2(r10, r00) and use r21 and r22 to obtain theta_x = atan2(r21, r22). the term c_y gets canceled in the division. But c_y has useful information that will decide if we need to add an offset of an offset +pi/2 or -pi/2 to asin(r20). 

Here is how  it work. Once we obtain theta_z we can substitute c_z  in r00 ((or s_z in r10)) and obtain c_y. We can follow a similar procedure starting from theta_x. With c_y at hand and knowing s_y from r22, we can use -tan2(s_y, c_y) to obtain theta_y to the full range of (-pi to +pi).

So, while I might be still missing something, I don't think the factor of two multiplication is justified.

**Reply by vladmandic** [reply]
that may be so - feel free to experiment with alternative and propose a pr if it yields better results.
